Generate docs for dotnet nuget commands.

rrelyea opened the pull request:
## Bug

Fixes: https://github.com/nuget/home/issues/9146
Regression: no

## Fix

Details: Helps generate documentation for dotnet commands.

## Testing/Validation

Tests Added: No
Reason for not adding tests:  Docs
Validation:  Will iterate with docs team and feedback from team.

Commit:
Generate docs for dotnet nuget commands.

rrelyea commented:
@karann-msft - can you give any feedback on the .md file in this change that will go to the dotnet cli docs?

nkolev92 commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/Commands.xml:
```diff
@@ -3,49 +3,82 @@
       <SingleValueOption Name="configfile" Help="Option_ConfigFile"/>
+      <Example Title="List configured sources from the current directory:" Command="dotnet nuget list source" />
+      <SeeAlso Title="Package source sections in NuGet.config files" Url="https://docs.microsoft.com/en-us/nuget/reference/nuget-config-file#package-source-sections" />
+      <SeeAlso Title="sources command (nuget.exe)" Url="https://docs.microsoft.com/en-us/nuget/reference/cli-reference/cli-ref-sources" />
```
If you are going to generate docs with these links included you need to generate different links not the culture specific ones: 

https://review.docs.microsoft.com/en-us/help/contribute/links-how-to?branch=master#links-to-articles-in-the-same-docset

zivkan replied:
although links to docs within the same product should be relative to the .md file, so that the docs validator can validate the URLs are valid. If you use URLs, it won't check the link is valid.

rrelyea replied:
these docs will live in SDK docset...so we need full url...but i'm removing the locale from the links. thanks. good catch.

nkolev92 replied:
I think the guidance is the not use the docs.microsoft.com part anyways.

rrelyea replied:
ah yes, i read Andy's link, but didn't read closely enough:
"For docsets that are also used in "offline books" (this includes Windows, .NET, ASP.NET, Visual Studio, C++, SQL Server, and NuGet)," ... use relative.

rrelyea replied:
fixed.

Commit:
Fix links to docs to be non-locale specific.

Commit:
for nuget, we are supposed to use relative uris, since they are used in offline books too.

Commit:
always run ttGen.ps1 when editing commands.xml or any .tt file.

loic-sharma commented on src/NuGet.Core/NuGet.CommandLine.XPlat/external/docs.md:
```diff
@@ -0,0 +1,364 @@
+
+---
+title: dotnet nuget add source command
+description: The `dotnet nuget add source` command will add a new source to your NuGet configuration files. This will enable finding packages at this additional source. 
```
What do you think of this?

```suggestion
description: The `dotnet nuget add source` command adds a new package source to your NuGet configuration files.
```

The two sentences are a little similar, hence why I merged them.

rrelyea replied:
yep. made the change.

loic-sharma commented on src/NuGet.Core/NuGet.CommandLine.XPlat/external/docs.md:
```diff
@@ -0,0 +1,364 @@
+
+  Password to be used when connecting to an authenticated source.
+
+- **`--store-Password-In-Clear-Text`**
```
Should this be lowercase to match the CLI option?

loic-sharma replied:
Same comment for `valid-Authentication-Types`

rrelyea replied:
yep...gonna send another commit. missed this comment.

loic-sharma commented on src/NuGet.Core/NuGet.CommandLine.XPlat/external/docs.md:
```diff
@@ -0,0 +1,364 @@
+
+- **`--valid-Authentication-Types`**
+
+  Comma-separated list of valid authentication types for this source. By default, all authentication types are valid. Example: basic,negotiate
```
What are all the available authentication types? Should we give a quick synopsis on what `basic` and `negotiate` means? Are there additional docs on these types that we can link? I would recommend wrapping the example in a code block: `basic,negotiate`

rrelyea replied:
will open doc issue to improve the details here...not blocking now.
tracking here: https://github.com/NuGet/Home/issues/9233

loic-sharma commented on src/NuGet.Core/NuGet.CommandLine.XPlat/external/docs.md:
```diff
@@ -0,0 +1,364 @@
+
+- [sources command (nuget.exe)](/nuget/reference/cli-reference/cli-ref-sources)
+
+}
```
I'm not familiar with this file's syntax - is this `}` intentional?

rrelyea replied:
nope. thx. fixing

rrelyea commented:
Offline feedback from Tom Dykstra (sdk doc lead)
    Yes, 6 files in our repo and then we can tweak the formatting in the PR that adds them. The format looks good, just a couple of issues I notice now -- 
- ms.date format requires leading zero
- our markdown linter requires a blank line before and after headings and code blocks, except the h1 goes  right after the yaml block.
- Acrolinx always advises to use present tense where possible -- "The ... command will update an existing source ..." --> "The ... command updates an existing source..."

Commit:
fix feedback in PR and offline with sdk docs lead

Commit:
lowercase all option string

Commit:
Provide file name and path info in concatenated file.

rrelyea commented:
@loic-sharma - can you please re-review.

loic-sharma commented on src/NuGet.Core/NuGet.CommandLine.XPlat/external/docs.md:
```diff
@@ -0,0 +1,372 @@
+
+- **`-u|--username`**
+
+  UserName to be used when connecting to an authenticated source.
```
Should this be `Username`? It’s a valid English word: https://www.merriam-webster.com/dictionary/username

loic-sharma replied:
There’s several occurrences of `UserName`, don’t forget to update those too if you change this.

rrelyea replied:
fixed in next commit. also fixed strings.resx (nuget.exe command)

loic-sharma commented on src/NuGet.Core/NuGet.CommandLine.XPlat/external/docs.md:
````diff
@@ -0,0 +1,372 @@
+- Enable a source with name of `mySource`:
+
+  ```dotnetcli
+  dotnet nuget remove source mySource
````
Is this the right example? Shouldn’t this be an update?

rrelyea replied:
fixed in next commit

loic-sharma commented on src/NuGet.Core/NuGet.CommandLine.XPlat/external/docs.md:
```diff
@@ -0,0 +1,372 @@
+
+- **`--format`**
+
+  Applies to the list action. Accepts two values: Detailed (the default) and Short.
```
What do you think of something like `The format of the list command output` to clarify this affects the output? Also can we wrap `Detailed` and `Short` in back ticks?

rrelyea replied:
fixed in next commit

Commit:
PR Feedback fixes (UserName -> Username, etc...)

nkolev92 commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/Commands.xml:
```diff
@@ -3,49 +3,82 @@
+      <Example Title="Add `nuget.org` as a source:" Command="dotnet nuget add source https://api.nuget.org/v3/index.json -n nuget.org" />
+      <Example Title="Add `c:\packages` as a local source:" Command="dotnet nuget add source c:\packages" />
+      <Example Title="Add a source that needs authentication:" Command="dotnet nuget add source https://someServer/myTeam -n myTeam -u myUsername -p myPassword --store-password-in-clear-text" />
+      <Example Title="Add a source that needs authentication (then go install credential provider):" Command="dotnet nuget add source https://azureartifacts.microsoft.com/myTeam -n myTeam" />
```
We should probably use contoso for the example urls.

rrelyea replied:
which one would you change? the ones with myTeam? can you make the suggesting more specific.

- nuget.org 
- c:\packages
- https://someserver/myTeam
- https://azureartifacts.microsoft.com/myTeam

nkolev92 replied:
azureartifacts.microsoft.com to contoso. 

Basically the base part of each url should be contoso.

rrelyea replied:
i was hoping to make that a real azure artifacts example. perhaps i need to tweak it to be more real.

nkolev92 replied:
I was under the impression that all examples in docs should be contoso to avoid confusion.

rrelyea replied:
I'm asking for feedback from the sdk lead who reviewed these docs:  https://github.com/dotnet/docs/pull/17401#issuecomment-606770645

rrelyea replied:
added more contoso in latest commit based on your feedback and then discussion with jonathan & tom (docs lead)

nkolev92 commented on build/ttGen.ps1:
```diff
@@ -0,0 +1,10 @@
+Write-Host "Regenerating all generated files from all .tt files."
```
move this to `scripts\`

nkolev92 replied:
When will this be run? Manually?

rrelyea replied:
moved to \scripts\utils and added some instructions.

nkolev92 reviewed (commented):
Just a few quick questions before approving.

How will the publishing, flow from here to the docs repo work?

Commit:
more instructions on scripts. change location of scripts.

rrelyea commented:
created initial PR in dotnet docs repo here: https://github.com/dotnet/docs/pull/17401

rrelyea commented:
> How will the publishing, flow from here to the docs repo work?

added some instructions in the different scripts/docs files. current process: 
- edit commands.xml and strings.resx to create commands/examples/etc.
- generate docs.md from docs.tt via scripts\util\ttgen.ps1
- split into separate files in your clone of dotnet/docs via scripts\util\docsPRGen.ps1
- do additional edits of TOC, etc.. in your fork, then create PR to dotnet/docs

Commit:
make encoding UTF8 so unicode check mark gets output. remove double blank lines after argument section.

Commit:
create file with ASCII (still have to handle unicode better)

Commit:
Fixes in order to improve formatting based on docs team feedback.

Commit:
per docs PR feedback - alphabetize options and fix validauthenticationtypes

Commit:
improve validauthtypes docs per docs team feedback

zivkan commented on scripts/utils/ttGen.ps1:
```diff
@@ -0,0 +1,12 @@
+        Write-Host t4.exe $ttFile
+        t4.exe "$ttFile"
+    }
+}
```
our repo's `.editorconfig` file specifies `insert_final_newline = true`, meaning your editor doesn't respect `.editorconfig` :(

rrelyea replied:
reading https://github.com/PowerShell/PowerShellEditorServices/issues/665 it seems that i needed to install the "editorconfig for vs code" extension. It then added an trailing newline. will commit change too.

zivkan commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/Commands.xml:
```diff
@@ -3,49 +3,82 @@
-    <Noun Name="source">
+    <Noun Name="source" Version="3.1.200"
+       Description="adds a new package source to your NuGet configuration files."
+       >
```
I don't think I've seen this formatting style before. Normally the `>` is on the end of the last line, similar to how `)` is normally at the end of the last parameter in multi-line expressions or method declarations in c#.  Our `NuGet.targets` file keeps the `>` or `/>` on the same line as the last attribute.

rrelyea replied:
fixed in next commit.

zivkan commented on src/NuGet.Core/NuGet.CommandLine.XPlat/external/docsHelpers.t4:
```diff
@@ -0,0 +1,19 @@
+        public bool IsArgument { get; set; }
+        public string ArgumentName { get; set; }
+    }
+#>
```
`insert_final_newline`?

rrelyea replied:
fixed in next commit

zivkan commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/Commands.xml:
```diff
@@ -3,49 +3,82 @@
+      <Example Title="Add `c:\packages` as a local source:" Command="dotnet nuget add source c:\packages" />
+      <Example Title="Add a source that needs authentication:" Command="dotnet nuget add source https://someServer/myTeam -n myTeam -u myUsername -p myPassword --store-password-in-clear-text" />
+      <Example Title="Add a source that needs authentication (then go install credential provider):" Command="dotnet nuget add source https://azureartifacts.microsoft.com/myTeam -n myTeam" />
+      <SeeAlso Title="Package source sections in NuGet.config files" Url="/nuget/reference/nuget-config-file#package-source-sections" />
```
By prefixing all these URLs with `/`, you're telling the docs validator to not validate the link.

https://github.com/NuGet/docs.microsoft.com-nuget#links

> Relative links to other pages in this docset should be in the form `../<folder>/<topic-file>.md` including the trailing .md.

rrelyea replied:
is dotnet and nuget the same docset?
the lint tool and the review by the sdk docs folks didn't raise an issue here.

zivkan replied:
the linter isn't smart enough to warn that a link starting with `/` is from the current docset, so no, it's not going to warn you.

But, I forgot that  the dotnet cli docs live in the dotnet docs repo and not ours. In that case, yes, we need to keep that `/` prefix.

rrelyea replied:
@zivkan - cool. can you approve then please?

zivkan reviewed (commented):
I agree with using contoso rather than realistic looking fake URLs.

Also, I think all the links to other docs pages needs to be fixed.

Commit:
PR feedback

Commit:
update URLs for examples to use more contoso

nkolev92 commented on scripts/utils/ttGen.ps1:
```diff
@@ -0,0 +1,12 @@
+Write-Host "Regenerating all generated files from all .tt files."
+Write-Host "(requires t4.exe from 'dotnet tool install --global dotnet-t4' or right click .tt files in VS and 'run custom tool')"
```
You can add a local tools manifest.
https://docs.microsoft.com/en-us/dotnet/core/tools/global-tools#install-a-local-tool

Then it will work automatically.

rrelyea replied:
more automatically...yes...still needs a dotnet tool restore -- or should we add that tool restore into configure.ps1?

nkolev92 replied:
yeah the idea would be to add it to configure.ps1.

rrelyea replied:
I've opened a tracking issue for that work: https://github.com/nuget/home/issues/9379

nkolev92 commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/Verbs.tt:
```diff
@@ -9,15 +9,27 @@
 <#
     string InitCaps(string input)
     {
+        if (input == null) return null;
```
style: 
```suggestion
        if (input == null) 
        {
            return null;
        }
```

Commit:
don't do 1 line if statement

Co-Authored-By: Nikolche Kolev <nikolce1kolev@gmail.com>

rrelyea commented:
https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=3619936 failed e2e and apex.
but same commit succeeded on those same two passes in this build: https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=3599215
